### PR TITLE
ASoC: SOF: Add protection for swidget->use_count concurent access in PCM and control path

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

The use_count of swidget is protected by PCM locking but the control path also checks it to decide to send a message to firmware or not (the swidget is present or not).
Since the kcontrol access is not protected by PCM locks there is a possibility that the swidget got freed up between the time the control code checks the use_count and sends the message (this is more likely in ipc3).

The second patch fixes this issue, the first one is something I stumbled upon when adding the mutex protection.

### CODE REVIEW

**Comment by ujfalusi** [comment]
@lyakh, FYI

File: sound/soc/sof/sof-audio.c [hunk 1]
```diff
@@ -182,13 +183,16 @@ int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
...
 core_put:
 	snd_sof_dsp_core_put(sdev, swidget->core);
 pipe_widget_free:
 	if (swidget->id != snd_soc_dapm_scheduler)
 		sof_widget_free(sdev, swidget->spipe->pipe_widget);
 use_count_dec:
-	swidget->use_count--;
+	if (!use_count_balanced)
+		swidget->use_count--;
+
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:208]: use_count_balanced isn't a great self-explanatory variable.

use_count_decremented maybe?

Comment (plbossart) [inline sound/soc/sof/sof-audio.c:208]: typo in commit message: jump[

File: sound/soc/sof/sof-audio.c [hunk 2]
```diff
@@ -116,8 +120,12 @@ int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 
 	trace_sof_widget_setup(swidget);
 
+	mutex_lock(&swidget->use_count_mutex);
+	swidget->use_count++;
+	mutex_unlock(&swidget->use_count_mutex);
+
 	/* widget already set up */
-	if (++swidget->use_count > 1)
+	if (swidget->use_count > 1)
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: does this work?

Assuming use_count is zero, you could have a race where two threads increment use_count in a serial manner, and then both test the use_count as 2 and both return without doing anything. I think you need to protect the rest, not just the incrementation of the variable.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/sof/sof-audio.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: yes, I think the both sof_widget-setup() and free() must be protected as a whole too

**Comment by plbossart** [comment]
@ujfalusi ping? is this a PR that will be updated? It's been 7 days since the feedback was provided.

File: sound/soc/sof/sof-audio.c (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: The threads which can increase/decrease are protected by higher level locking in PCM level, we need to protect this against the checks happening outside of the PCM lock, like the kcontrol access.

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: Right, hrm, there might be concurrent PCM streams touching the same widget? That might be an issue, but that issue already exists, no? The chance is slim, but in theory it could happen, the `use_count` is not atomic...

File: sound/soc/sof/sof-audio.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c:208]: OK

File: sound/soc/sof/sof-audio.c (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: Blindly protecting these function as whole results (when running aplay):
```
[13115.199981] ================================================
[13115.199984] WARNING: lock held when returning to user space!
[13115.199988] 6.1.0-rc6+ #841 Not tainted
[13115.199993] ------------------------------------------------
[13115.199995] aplay/13597 is leaving the kernel with locks still held!
[13115.199997] 2 locks held by aplay/13597:
[13115.200000]  #0: ffff9a1304375c88 (&swidget->use_count_mutex){....}-{3:3}, at: sof_widget_setup+0x84/0x3a0 [snd_sof]
[13115.200038]  #1: ffff9a1304374e88 (&swidget->use_count_mutex){....}-{3:3}, at: sof_widget_setup+0x84/0x3a0 [snd_sof]
```
Needs to be narrowed down or to protect the increment/decrement + the test.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: I suppose protecting the use_count during control IO is good enough then

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/sof-audio.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: Can we not have multiple controls using the same widget? I think this needs a lot more explanations on what agents will simultaneously use swidget->use_count

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: yes, we can but the mutex would protect all the get/put ops for all the kcontrols right?

Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: the other question I would have is that if we only want to protect the kcontrol path, and the incrementation and decrementation of a variable is really protected by a PCM mutex, isn't a spin_lock good enough and the mutex overkill?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: why do you think that @plbossart. I'd think its the other way around.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: besides a spinlock cannot be used for the purpose of this PR at all because we'd have to sleep while waiting for a reply for set_get_data IPC and a spinlock is meant to be atomic right?

**Comment by ujfalusi** [comment]
Changes since v1:
- Protect the entire sof_sidget_setup/free functionality
- rename the mutex to a more generic (`setup_mutex`) now that it is protecting the setup/free (free == **un**setup ;) )
- Use `use_count_decremented` as flag for use_count balancing flag
- typoes fixed

File: sound/soc/sof/ipc3-control.c [hunk 3]
```diff
@@ -647,7 +659,7 @@ static int sof_ipc3_widget_kcontrol_setup(struct snd_sof_dev *sdev,
 	list_for_each_entry(scontrol, &sdev->kcontrol_list, list)
 		if (scontrol->comp_id == swidget->comp_id) {
 			/* set kcontrol data in DSP */
-			ret = sof_ipc3_set_get_kcontrol_data(scontrol, true);
+			ret = sof_ipc3_set_get_kcontrol_data(scontrol, true, false);
```
Comment (plbossart) [inline sound/soc/sof/ipc3-control.c:664]: can we add a deplock_assert to make sure it's fine to have no lock here?

Comment (ujfalusi) [reply sound/soc/sof/ipc3-control.c:664]: By definition the `widget_kcontrol_setup` op is called from `sof_widget_setup()` which is the locked, but yes, we could try.

File: sound/soc/sof/sof-audio.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: you're right @ranj063 not sure what I was thinking

File: sound/soc/sof/ipc3-control.c (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc3-control.c:664]: @plbossart, if you insist I can add
```c
WARN_ON_ONCE(!mutex_is_locked(&swidget->setup_mutex));
```
here and the ipc4 callback as well, but it will never catch anything as this is always happening at widget setup.

Comment (plbossart) [reply sound/soc/sof/ipc3-control.c:664]: I was thinking lockdep_assert_held(&swidget->setup_mutex);

File: sound/soc/sof/ipc3-control.c [hunk 4]
```diff
@@ -40,8 +42,11 @@ static int sof_ipc3_set_get_kcontrol_data(struct snd_sof_control *scontrol, bool
 	 * would always be > 0 in this case. For the others, just return the cached value if the
 	 * widget is not set up.
 	 */
+	if (lock)
```
Comment (ranj063) [inline sound/soc/sof/ipc3-control.c:40]: ths comment above is in the wrong place now

File: sound/soc/sof/ipc3-control.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/ipc3-control.c:664]: @plbossart @ujfalusi I think what we could do instead is split the sof_ipc3_set_get_kcontrol_data() into unlock and regular functions and call the unlock version from sof_ipc3_widget_kcontrol_setup() and then add the lockdep_assert() here in

File: sound/soc/sof/ipc3-control.c (hunk 4 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc3-control.c:40]: OK

File: sound/soc/sof/ipc3-control.c (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc3-control.c:664]: > @plbossart @ujfalusi I think what we could do instead is split the sof_ipc3_set_get_kcontrol_data() into unlock and regular functions and call the unlock version from sof_ipc3_widget_kcontrol_setup() and then add the lockdep_assert() here in

@ranj063, I did tried to split the *-control.c functions as well but it turned out to be not so pretty. We have common code to find the swidget and it will create a web of calls which is hard to follow. The additional flag parameter was the cleanest way to do this.

Comment (ujfalusi) [reply sound/soc/sof/ipc3-control.c:664]: > I was thinking lockdep_assert_held(&swidget->setup_mutex);

I see, that could be added with the similar note that it is just a NOP as the setup is only going to happen during the setup.

**Comment by ujfalusi** [comment]
Changes since v3:
- Add `lockdep_assert_held(&swidget->setup_mutex);` when the set_get_kcontrol_data is called with lock=false to make sure that the lock is held
- move the comments for the use count checking to correct place.

File: sound/soc/sof/ipc4-control.c [hunk 5]
```diff
@@ -35,23 +37,34 @@ static int sof_ipc4_set_get_kcontrol_data(struct snd_sof_control *scontrol, bool
 		return -ENOENT;
 	}
 
+	if (lock)
+		mutex_lock(&swidget->setup_mutex);
+	else
```
Comment (ranj063) [inline sound/soc/sof/ipc4-control.c:42]: isnt the else redundant here?

Comment (plbossart) [reply sound/soc/sof/ipc4-control.c:42]: well one could argue that if you test that the mutex_lock did work it's also redundant...

**Review by plbossart — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]
